cli/mender-artifact/copy.go

```diff
@@ -0,0 +1,243 @@
+	// and thus we spare a fair few seconds not repacking it.
+	if repack {
+		wpc, ok := w.(PartitionPacker)
+		if !ok {
+			return cli.NewExitError("critical implementation error. Image cannot be repacked", 1)
+		}
+		if err = wpc.Repack(); err != nil {
+			return cli.NewExitError(fmt.Sprintf("failed to repack image: %v", err), 1)
+		}
+
+	}
+	if err = w.Close(); err != nil {
+		return err
+	}
+	return r.Close()
+}
+
+func Append(c *cli.Context) error {
+	fmt.Println("Hello")
+
+	var r io.ReadCloser
+	// var w io.WriteCloser
+	var rwcp PartitionReadWriteClosePacker
+	var appendint = 3 // positive infinity
+	var readindex = 0
+	var err error
+
+	for i, arg := range c.Args() {
+		if arg == "-" {
+			appendint = i
+			break
+		}
+	}
+
+	isimg := regexp.MustCompile(`\.(mender|sdimg)`)
+
+	switch {
+
+	case isimg.MatchString(c.Args().First()):
+		rwcp, err = NewPartitionFile(c.Args().First(), c.String("key"))
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		readindex = 1
+
+	case isimg.MatchString(c.Args().Get(1)):
+		rwcp, err = NewPartitionFile(c.Args().Get(1), c.String("key"))
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		readindex = 0
+
+	default:
+		return cli.NewExitError("a mender artifact or sdimg needs to be specified", 1)
+
+	}
+
+	buf := bytes.NewBuffer([]byte{})
+
+	switch appendint {
+	// prepend
+	case 0:
+		// write the stdin input to the buf first
+		if _, err = io.Copy(buf, os.Stdin); err != nil {
+			return cli.NewExitError(err.Error(), 1)
+		}
+
+		// then the partition-file
+		if _, err = io.Copy(buf, rwcp); err != nil {
+			return cli.NewExitError(err.Error(), 1)
+		}
+
+	// append
+	case 1:
+		// write image files' file to the buf first
+		if _, err = io.Copy(buf, rwcp); err != nil {
+			return cli.NewExitError(err.Error(), 1)
+		}
+		// then from stdin
+		if _, err = io.Copy(buf, os.Stdin); err != nil {
+			return cli.NewExitError(err.Error(), 1)
+		}
+
+	// read from input file
+	case 3:
+		// write image files' file to the buf first
+		if _, err = io.Copy(buf, rwcp); err != nil {
+			return cli.NewExitError(err.Error(), 1)
+		}
+		r, err = os.Open(c.Args().Get(readindex))
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		if _, err = io.Copy(buf, r); err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+
+	}
+
+	// then write it all back out into the image
+	if _, err = io.Copy(rwcp, buf); err != nil {
+		return cli.NewExitError(err.Error(), 1)
+	}
+
+	// finally repack the image
+	return rwcp.Repack()
+}
+
+// enumerate cli-options
+const (
+	copyin = iota
+	copyinstdin
+	copyout
+	parseError
+	argerror
+	criterror
+)
+
+func parseCLIOptions(c *cli.Context) int {
+
+	finfo, err := os.Stdin.Stat()
+	if err != nil {
+		return criterror
+	}
+
+	isimg := regexp.MustCompile(`\.(mender|sdimg)`)
+
+	// no data on stdin
+	if finfo.Mode()&os.ModeNamedPipe == 0 {
```

Comment:
I don't understand this, why are we checking for pipe status? Maybe a question for @oleorhagen.

Reply:
It is to see wether there is any data available on stdin using finfo see #220